### PR TITLE
COCOS-132 - Add progress bar for algo and data uploads

### LINKED ISSUE #132: Feature: Add progress bar for algo and data uploads
### Is your feature request related to a problem? Please describe.

User needs to be able to see the percentage of file uploads on cli for visibilirt

### Describe the feature you are requesting, as well as the possible use case(s) for it.

User needs to be able to see the percentage of file uploads on cli for visibility

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by danko-miladinovic**
# What type of PR is this?

This is a feature because it adds the following functionality:
- Adds a progress bar so that the user can see how much of the algorithm and data is uploaded to the Agent.
## What does this do?
This PR adds a `ProgressBar` struct that is used to draw the progress bar onto the terminal. The example of the progress bar can be seen here:
`Uploading data... 25% [=====>                        ]`

## Which issue(s) does this PR fix/relate to?

- Resolves #132 

## Have you included tests for your changes?

This PR can be tested with existing manual tests.
## Did you document any new/modified feature?
This feature does not need to be documented, because it is a standard feature.

### CODE REVIEW

File: pkg/progressbar/progressbar.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	if _, err := (*stream).CloseAndRecv(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *ProgressBar) SendData(description string, buffer *bytes.Buffer, stream *agent.AgentService_DataClient) error {
```
Comment (drasko) [inline pkg/progressbar/progressbar.go:80]: There is a lot of repeating code in these two functions. Can you make one common function with parameter algo/data to unify this?

File: pkg/sdk/agent.go [hunk 2]
```diff
@@ -90,28 +77,14 @@ func (sdk *agentSDK) Data(ctx context.Context, dataset agent.Dataset, privKey *r
...
-		}
-
-		err = stream.Send(&agent.DataRequest{Dataset: buf[:n]})
-		if err != nil {
-			return err
-		}
-	}
-
-	if _, err := stream.CloseAndRecv(); err != nil {
+	progressbar := progressbar.New()
```
Comment (drasko) [inline pkg/sdk/agent.go:85]: Do you need to create a new one, or you can reuse/reset old one if not used anymore?

**Review by drasko — CHANGES_REQUESTED** [review]

File: pkg/progressbar/progressbar.go (hunk 1 above)
Comment (SammyOina) [reply pkg/progressbar/progressbar.go:80]: @danko-miladinovic consider something like this 
```go
type StreamSender interface {
    Send(interface{}) error
    CloseAndRecv() (interface{}, error)
}

func (p *ProgressBar) sendData(description string, buffer *bytes.Buffer, stream StreamSender, createRequest func([]byte) interface{}) error {
    p.reset(description, buffer.Len())

    buf := make([]byte, bufferSize)

    for {
        n, err := buffer.Read(buf)
        if err == io.EOF {
            if _, err := io.WriteString(os.Stdout, "\n"); err != nil {
                return err
            }
            break
        }
        if err != nil {
            return err
        }

        p.updateProgress(n)

        if err := stream.Send(createRequest(buf[:n])); err != nil {
            return err
        }

        if err := p.renderProgressBar(); err != nil {
            return err
        }
    }

    _, err := stream.CloseAndRecv()
    return err
}

func (p *ProgressBar) reset(description string, totalBytes int) {
    p.currentUploadedBytes = 0
    p.currentUploadPercentage = 0
    p.numberOfBytes = totalBytes
    p.description = description
}

func (p *ProgressBar) updateProgress(bytesRead int) {
    if p.currentUploadedBytes < p.numberOfBytes {
        p.currentUploadedBytes += bytesRead
        p.currentUploadPercentage = float64(p.currentUploadedBytes) * 100.0 / float64(p.numberOfBytes)
    }
}

func (p *ProgressBar) SendAlgorithm(description string, buffer *bytes.Buffer, stream *agent.AgentService_AlgoClient) error {
    return p.sendData(description, buffer, *stream, func(data []byte) interface{} {
        return &agent.AlgoRequest{Algorithm: data}
    })
}

func (p *ProgressBar) SendData(description string, buffer *bytes.Buffer, stream *agent.AgentService_DataClient) error {
    return p.sendData(description, buffer, *stream, func(data []byte) interface{} {
        return &agent.DataRequest{Dataset: data}
    })
}
```

Comment (drasko) [inline pkg/progressbar/progressbar.go:80]: Do we need to export this function?

**Review by drasko — CHANGES_REQUESTED** [review]

File: pkg/sdk/agent.go (hunk 2 above)
Comment (danko-miladinovic) [reply pkg/sdk/agent.go:85]: Yes, progress bar is used in Agent SDK so we need to create a new one every time the users uses the CLI to send data and algo.

File: pkg/progressbar/progressbar.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+const (
+	progressBarDots = "... "
+	leftBracket     = "["
+	rightBracket    = "]"
+	head            = ">"
+	body            = "="
+	bodyPadding     = "."
+	bufferSize      = 1024 * 1024
+)
+
```
[the remark this answers is not in the document]
Comment (danko-miladinovic) [reply pkg/progressbar/progressbar.go:25]: Done

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**